Ensure generator picks up ApplicationSerializer

richmolj opened the pull request:
#### Purpose

Fixes https://github.com/rails-api/active_model_serializers/issues/1890

#### Changes

Look at filesystem instead of using `defined?`

#### Caveats

If the file was there but the class was not named correctly, I think you'd get an unfriendly error.

#### Related GitHub issues

https://github.com/rails-api/active_model_serializers/issues/1890

#### Additional helpful information

ApplicationSerializer could exist, but not be loaded. So, we check
existence by looking at the filesystem instead of defined?.

richmolj commented:
@bf4 attempted the fix you suggest in [this comment] but didn't work. Not sure if there is a better filesystem stub for this, I'm used to rspec.

richmolj commented:
Hm it actually just dawned on me if you *don't* have a namespace/application_serializer but do have a non-namespaced one, we won't fall back to the non-namespaced one. Do we care about this use case / what's the correct behavior?

bf4 commented on lib/generators/rails/serializer_generator.rb:
```diff
@@ -23,12 +23,14 @@ def association_names
+          namespace = class_path.map(&:camelize).join('::')
+          application_serializer = [namespace, 'ApplicationSerializer']
+            .reject(&:blank?).join('::')
+          path = File.join('app/serializers', class_path, "application_serializer.rb")
```
since this is only relevant in a Rails app, I'd rather rely on Rails autoloading for a more robust support and less code on us.

i.e.

```ruby
  return parent if (parent = options[:parent])
  ApplicationSerializer.name
rescue NameError
  'ActiveModel::Serializer'
end
```

bf4 replied:
(or use `safe_constantize` or different logic, but same idea)

richmolj replied:
Yep, I gave that a shot based on your original comment. Does not seem to work, I believe because these files aren't `required` at this time.

bf4 commented on test/generators/serializer_generator_test.rb:
```diff
@@ -20,11 +20,10 @@ def test_generates_a_namespaced_serializer
-    run_generator
+    File.stub(:exists?, true) do
+      run_generator
+    end
```
we might want to make this an isolated test...

richmolj commented:
@bf4 actually you were right, [this](https://github.com/rails-api/active_model_serializers/issues/1890#issuecomment-242818408) won't work but `safe_constantize` will. Removed the namespace stuff, added changelog.

New test had to get a little wonky though.

bf4 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ Breaking changes:
 Fixes:
 
 - [#1887](https://github.com/rails-api/active_model_serializers/pull/1887) Make the comment reflect what the function does (@johnnymo87)
+- [#1890](https://github.com/rails-api/active_model_serializers/issues/1890) Ensure generator Inherits from ApplicationSerializer when available (@richmolj)
```
`s/Inherits/inherits`

bf4 commented on test/generators/serializer_generator_test.rb:
```diff
@@ -54,4 +53,21 @@ def test_with_no_attributes_does_not_add_extra_space
+    String.class_eval do
+      alias_method :safe_constantize, :old
+    end
+  end
```
Or just

```ruby
class << test_string'
  define_method(:safe_constantize) do 
    Class if self == expected
  end
end
```

bf4 replied:
Or you could `undef` in the ensure

richmolj replied:
I don't think the `class << "ApplicationSerializer"` will work - gave it a shot and failed. Wouldn't that only patch the singleton of that particular *instance* of String?

NullVoxPopuli commented on test/generators/serializer_generator_test.rb:
```diff
@@ -20,11 +20,10 @@ def test_generates_a_namespaced_serializer
-    assert_file 'app/serializers/account_serializer.rb', /class AccountSerializer < ApplicationSerializer/
-  ensure
-    Object.send :remove_const, :ApplicationSerializer
+    stub_safe_constantize(expected: 'ApplicationSerializer') do
```
what's a failure look like when `self != expected` (inverse of line 64)?

richmolj replied:
`safe_constantize` would return `nil`, meaning the generator would (incorrectly) inherit from `ActiveModel::Serializer` instead of `ApplicationSerializer`:

```
  1) Failure:
SerializerGeneratorTest#test_uses_application_serializer_if_one_exists [/Users/lrichmond1/code/active_model_serializers/test/generators/serializer_generator_test.rb:25]:
Expected /class AccountSerializer < ApplicationSerializer/ to match "class AccountSerializer < ActiveModel::Serializer\n  attributes :id, :name, :description\n  has_one :business\nend\n".
```

bf4 commented:
@richmolj LGTM, you wanna merge?
